### PR TITLE
solana-program: `StakeHistorySyscall` and `get_entry`

### PR DESCRIPTION
**Pull request by 2501babe**
#### Problem
to support the native to bpf conversion of the stake program, we need a way to get stake history in instructions that dont pass in the account data, to replace the usage of `SysvarCache` from `InvokeContext` for that purpose

#### Summary of Changes
in #560 and #1307 we implemented a syscall to get the necessary sysvar data. this pr provides an interface over that syscall usable by bpf programs

an object `StakeHistorySysvar` is constructed with the current `Clock.epoch`, which allows it to calculate offsets to retrieve arbitrary stake history entries in O(1), which is does via the `sol_get_sysvar` syscall. we also introduce a trait, `StakeHistoryGetEntry`, which provides the method `get_entry`, and change important functions on `Stake` and `Delegation` to use implementers of this trait interchangebly, so that all existing callers of the relevant functions remain unchanged

after we convert the native programs, there will be very few remaining callers of `get` and we can deprecate it in favor of `get_entry`

### CODE REVIEW

File: sdk/program/src/stake_history.rs
Comment (2501babe) [inline sdk/program/src/stake_history.rs]: @joncinque curious for your thoughts on whether its semantically reasonable to return `None` for error conditions like this

the existing `StakeHistory` is just a vector, which `get()` searches for a given epoch, and if the epoch isnt there, it returns `None`. the important consideration is `None` is a _success_ condition: the `StakeHistoryEntry` was not found, so the `Delegation` function assumes that the stake is fullly active or inactive

the issue here is the `None` returns inside the `match` are _failure_ conditions. they should probably never be hit. the specific cases are things like "stake history became sparse so the offsets are wrong" (probably impossible), "stake history is shorter than it should be unexpectedly" (probably impossible), "stake history is not loaded in syscar cache" (almost certainly impossible), things like that

but i just wanted to get a second opinion, do you think its acceptable to return `None` for these? the alternative is returning `Result<Option<_>, _>` and modifying the `Delegation` functions such as `activating_and_deactivating()` to handle that... which im wary of touching that logic, and also its not clear to me what we would do, other than abort execution or just behave as if we got `None`

(i realized after pushing this that the bailing returns at the start of `get_entry()` are all fine, because it does the same thing as `get()`, so the stuff inside the `match` is all that im concerned about)

Comment (joncinque) [reply sdk/program/src/stake_history.rs]: This is a tricky one, since there's two things going on.

One the one hand, the code clearly checks that you're within the right bounds, so in normal circumstances, it would be better to `assert!` since otherwise there would be a programmer error.

On the other hand, things can change in the runtime, so the syscall could possibly fail in the future if there's ever a breaking change on the sysvar. This is unlikely, but not impossible. In those cases, it would be better to give the program a chance to notice the problem and recover rather than crash right away.

`Result<Option<...>>` seems a bit overkill, so all that to say, I agree with this approach.

Comment (2501babe) [reply sdk/program/src/stake_history.rs]: well the good news is the errors for the syscall are specified in extreme detail, because of the whole simd process. it ended up being a bit strange though because there are a couple cases where the sycall fails successfully (returns a non-`SUCCESS` error code), namely reading past the end of account data, or the sysvar not being available/supported (ie it was never populated in the sysvar cache, but in theory this would never happen)

in other words i dont think an update to the sysvar can get around needing to update the syscall and its interface. but ok if youre happy with the `None` returns im happy too

**Comment by 2501babe** [comment]
alright this is done now, i have to write tests before i merge it but im going to do some stake program things first. cc @buffalojoec if you wanted to copy for slot hashes

**Comment by buffalojoec** [comment]
@2501babe is there some formal process for articulating that this new API can't be used until the feature is activated? It will be available in 2.0 but until any cluster activates the feature it's always going to return `None`. Although it will succeed with the test validator.

**Comment by 2501babe** [comment]
as best im aware theres not. actually i guess it wont be relevant for awhile, mainnet isnt on 1.18 yet so the "no bpf conversions until after breakpoint" constraint is probably not the blocker, 2.0 feature activations will be

it seems harmless to merge this into 2.0 because nothing will be able to use it anyway, until the syscall activates it will panic. but maybe it would be more polite to wait until feature activation and then backport it into the next 2.0.x version 

ill ask releng when it becomes relevant

**Comment by joncinque** [comment]
Note that the BPF loader will prevent any deployments of programs that include a syscall that hasn't been enabled. I hunted around and I think it finally happens here: https://github.com/solana-labs/rbpf/blob/0270240454d47df3a81063b96248f266d62bf1f0/src/elf.rs#L1093

which is called from here https://github.com/anza-xyz/agave/blob/00d212e4abe9c8c736c6affd45c5b494711bc2a1/programs/bpf_loader/src/lib.rs#L128-L131

So that means you can safely merge this in -- people's deployments will just fail when they go to mainnet

**Comment by 2501babe** [comment]
```
---- sysvar::stake_history::tests::test_stake_history_get_entry stdout ----
HANA current: 514, newest: 513, oldest: 2, target: 2, delta: 511, offset: 16360, e&s: 32, u64: 8
HANA solgetsv stub, offset: 16360, length: 32, data len: 16392
HANA current: 514, newest: 513, oldest: 2, target: 3, delta: 510, offset: 16328, e&s: 32, u64: 8
HANA solgetsv stub, offset: 16328, length: 32, data len: 16392
HANA current: 514, newest: 513, oldest: 2, target: 4, delta: 509, offset: 16296, e&s: 32, u64: 8
HANA solgetsv stub, offset: 16296, length: 32, data len: 16392
HANA current: 514, newest: 513, oldest: 2, target: 5, delta: 508, offset: 16264, e&s: 32, u64: 8
HANA solgetsv stub, offset: 16264, length: 32, data len: 16392
HANA current: 514, newest: 513, oldest: 2, target: 6, delta: 507, offset: 16232, e&s: 32, u64: 8
HANA solgetsv stub, offset: 16232, length: 32, data len: 16392
HANA current: 514, newest: 513, oldest: 2, target: 7, delta: 506, offset: 16200, e&s: 32, u64: 8
HANA solgetsv stub, offset: 16200, length: 32, data len: 16392
HANA current: 514, newest: 513, oldest: 2, target: 8, delta: 505, offset: 16168, e&s: 32, u64: 8
HANA solgetsv stub, offset: 16168, length: 32, data len: 16392
HANA current: 514, newest: 513, oldest: 2, target: 9, delta: 504, offset: 16136, e&s: 32, u64: 8
thread 'sysvar::stake_history::tests::test_stake_history_get_entry' panicked at sdk/program/src/sysvar/stake_history.rs:125:17:
assertion `left == right` failed
  left: 11577965315787831468
 right: 9
```

relevant code:
```rust
        match result {
            Ok(()) => {
                let (entry_epoch, entry) =
                    bincode::deserialize::<(Epoch, StakeHistoryEntry)>(&entry_buf).ok()?;

                // this would only fail if stake history skipped an epoch or the binary format of the sysvar changed
                assert_eq!(entry_epoch, target_epoch);

                Some(entry)
            }       
            _ => None,
        }
```

the test creates a full stake history and then uses it with a syscall stub, and i was accidentally filling out stake history entries like this:
```rust
        let unique_entry_for_epoch = |epoch: u64| StakeHistoryEntry {
            activating: epoch % 2,
            deactivating: epoch % 3,
            effective: epoch % 5,
        };
```

~~and i meant to use `*` obviously, since i just wanted locally unique numbers to detect offset issues conveniently. but the fact that using `%` gives me this issue makes me wonder how it could even happen, since it doesnt occur for me locally when i run the test, but this test fails in the `coverage` job which depends on docker and buildkite etc...~~

~~the best i can think of is that the round trip through bincode, with enough zero values in the right places, makes it truncate something. and in practice this should never matter (if that is the cause) because none of these numbers are ever going to be zero~~

but it bothers me that i cant figure out the cause and im considering replacing the `assert_eq` with a `return None` just in case so it could fail gracefully. but if anyone has thoughts _why_ it would do this id be very interested...

edit: well that theory was wrong. my gut says this is an issue with the unsafe slice copy in the stub rather than `get_entry` itself now tho

**Comment by 2501babe** [comment]
well this gets stranger. heres a good execution of the loop:
```
HANA current: 514, newest: 513, oldest: 2, target: 6, delta: 507, offset: 16232, e&s: 32, u64: 8
HANA creating buffer
HANA calling syscall
HANA stub
HANA serializing data
HANA serialization ok? true
HANA solgetsv stub, offset: 16232, length: 32, data len: 16392
HANA copied!! 16232..16264
     data: [6, 0, 0, 0, 0, 0, 0, 0, 30, 0, 0, 0, 0, 0, 0, 0, 12, 0, 0, 0, 0, 0, 0, 0, 18, 0, 0, 0, 0, 0, 0, 0]
    slice: [6, 0, 0, 0, 0, 0, 0, 0, 30, 0, 0, 0, 0, 0, 0, 0, 12, 0, 0, 0, 0, 0, 0, 0, 18, 0, 0, 0, 0, 0, 0, 0]
HANA done! syscall
```

and heres the next iteration, which fails:
```
HANA current: 514, newest: 513, oldest: 2, target: 7, delta: 506, offset: 16200, e&s: 32, u64: 8
HANA creating buffer
HANA calling syscall
HANA stub
HANA done! syscall

thread 'sysvar::stake_history::tests::test_stake_history_get_entry' panicked at sdk/program/src/sysvar/stake_history.rs:128:17:
assertion `left == right` failed
  left: 17871374346716812888
 right: 7
```

whats weird about this is `HANA stub` and `HANA done! syscall` prints are located before and after the `crate::program_stubs::sol_get_sysvar()` call and the additional prints in the success case happen inside the stub that the test overrides the default stub with, which means... in the failure case it isnt calling the stub at all?

but if that were the case, then `left` should be zero because the memory is zero-initialized, so _something_ has to be happening. i think im out of ideas on how to proceed, especially since i cant reproduce this locally

**Comment by joncinque** [comment]
I can reproduce this locally by running all the tests -- there's a collision with the other test that's setting the syscall stubs at https://github.com/anza-xyz/agave/blob/83527d95777b21cdd155a6293ca4fdc740eda7a6/sdk/program/src/sysvar/slot_hashes.rs#L191-L192

When I ignore that test, everything is fine.

If we want both tests to set syscall stubs, we might want to have a common set of stubs protected by a `Once` rather than having both tests set their own

**Comment by 2501babe** [comment]
omg tysm. i thought rust tests ran in their own individually isolated memory so i assumed it had to be something with the ci environment. an easy fix knowing the cause tho

**Comment by buffalojoec** [comment]
Oops, my bad! I hadn't had a chance to look into this yet, but glad @joncinque found it quickly.

File: sdk/program/src/stake_history.rs
Comment (joncinque) [inline sdk/program/src/stake_history.rs]: I understand the need, but we should avoid using `Arc` in solana-program. Can you use `stake_history.as_ref()` in `programs/stake/src/stake_state.rs` and remove this? It worked when I tried it

File: sdk/program/src/sysvar/stake_history.rs
Comment (joncinque) [inline sdk/program/src/sysvar/stake_history.rs]: Sorry if this is a dumb question or I'm missing something obvious, but how does this work for testing? With something like program-test, do we always need to advance the clock for it to work? Or if I'm using solana-test-validator, will the stake program fail during the first epoch?

I see that we avoid using `get_entry` for things like `current_epoch == target_epoch`, so does that help us avoid ever touching the error path?

File: sdk/program/src/sysvar/stake_history.rs
Comment (joncinque) [inline sdk/program/src/sysvar/stake_history.rs:160]: hehe, clever

**Review by joncinque — COMMENTED** [review]
Looks great! Just the one bit about `Arc`, but otherwise this is good to go

File: sdk/program/src/sysvar/stake_history.rs
Comment (2501babe) [reply sdk/program/src/sysvar/stake_history.rs]: this is because i thought the first epoch was 1 rather than 0... ill remove it and see how it affects the math

**Review by joncinque — APPROVED** [review]

**Pull request merged**
